### DISCUSSION TITLE
Allow HTML responses from loaders to bypass default component exports

### DISCUSSION THREAD

**Discussion by wrporter** [remix-run/remix#5835] [Proposals]
I am migrating a massive site from WordPress to a new headless CMS. For security reasons, we store the legacy site in an S3 bucket rather than running a hot WordPress instance. While we migrate onto our new CMS and Remix system, we support legacy pages via a proxy to S3 if there are no pages in the new CMS or routes specified in Remix for a specific URL path. We manage this proxy in `entry.server.tsx`, but this is far from ideal as we aren't able to have as much control with the Remix data as we'd like. 

I propose a new feature where if a route loader returns an HTML response, it bypasses the default component export. In our case, it would be to proxy the request to the legacy S3 page if it is not found in the new CMS yet. HTML responses can be returned by loaders when there is no default component export, but it is not possible as soon as the default export is added. We want the default export to exist for rendering pages that do exist in our new CMS.

**Comment by sergiodxa** [comment]
I just found a way you could do this.

Let's say you have a route `routes/test.tsx` that has the following code:

```tsx
import { json, useLoaderData } from "remix";
import { useHydrated } from "remix-utils";

export function loader() {
  return json({
    type: "html",
    html: "<p>hola mundo</p>",
  });
}

export default function Component() {
  let { type } = useLoaderData<{ type: "html" | "json" }>();
  let hydrated = useHydrated();
  if (type === "html") {
    if (hydrated) window.location.reload();
    else return null;
  }
  return <p>hello world</p>;
}

```

The route return a json with a type and html fields, then in the component it checks the type, if it's HTML and the app is already hydrated calls `location.reload()`, otherwise return null.

Then in your `entry.server` you can add
```ts
let routeData = Object.values(remixContext.routeData).find((routeData) => {
  return routeData?.type === "html";
});

if (routeData) return html(routeData.html);
```

> **Note**: the `html` function comes from remix-utils

The code in `entry.server` will find the first route that returned `type === "html"`, and instead of rendering RemixServer it will send an immediate response with the `html` string.

If the user clicks a `Link` to this route, and the loader returns `type === "html"` the `location.reload()` will make the browser do a new document request and that way get the HTML directly.

**Reply by wrporter** [reply]
@sergiodxa Thank you for taking a look into this! I really appreciate you taking the time. This is the only major pain point I've experienced with Remix since adopting it 1 year ago.

It looks like what you came up with might work for us. But there are more complex details we'd have to consider with this half-baked approach. For example, there are many headers that come back in the response from S3 that we want to just send right back out to the client. For example:

```jsx
export function loader({ request }) {
  const requestUrl = new URL(request.url);
  const cmsPage = await getCmsPage(requestUrl);

  if (!cmsPage) {
    // The page does not exist in the CMS, proxy to legacy page in S3.
    // Straight proxy of response details from S3 to the client.
    return await fetch(`${s3BaseUrl}${requestUrl.pathname}`);
  }

  // The page exists in the new CMS, let's render it with Remix
  return json(cmsPage);
}

export default function Page() {
  const cmsPage = useLoaderData();
  // Render CMS page type accordingly.
  return <div>{cmsPage}</div>;
}
```
